### ISSUE TITLE
Analyzing/optimzing timing and speed of data acquisition

### ISSUE THREAD

**Issue by xjjak** [xjjak/LapCal#1]
One full reading of all 6 sensors takes *x* ms. (*How long does it actually take?*) Can this be improved?

**1 Idea**: Resetting all FIFO buffers of the DMPs before starting to read each of them.
- Does the first buffer fill up before it is watched again?
- How many buffers could be cleared ahead of time?

**Comment by xjjak** [comment]
The results of the first analysis:
```
Avg. Mux switch (3969): 181
Avg. Buf reset (3968): 425
Avg. Buf wait (3968): 4767
Avg. Buf read (3968): 1288
```
The values are in microseconds and are averaged across the number of readings also given in the parentheses.

The next evaluation includes a larger dataset and an additional measurement of the time taken for the whole reading process:
```
Avg. Mux switch (12956): 175
Avg. Buf reset (12956): 426
Avg. Buf wait (12956): 4673
Avg. Buf read (12956): 1282
Avg. reading of all sensors (2592): 59626
```

After implementing a first optimization, resetting the FIFO buffers of all sensors, before starting to read from the first one, an improvement of `~7ms` for the whole reading process could be noted:
```
Avg. Mux switch (5173): 142
Avg. Buf reset (5174): 0
Avg. Buf wait (5174): 878
Avg. Buf read (5173): 1274
Avg. reading of all sensors (1035): 52549
```
Here, it can also be seen, that this decreased the duration of the avg. waiting period for the buffer to fill up, significantly from `~4700` to `~900`. The overall improvement is much less though, as the sensors' buffers still have to be reset beforehand and be switched to on the multiplexer. 

Although it makes little sense, resetting the last sensor and going from "right-to-left", as small improvement of `~200` could be noted.
```
Avg. Mux switch (5525): 111
Avg. Buf reset (5525): 0
Avg. Buf wait (5525): 957
Avg. Buf read (5525): 1278
Avg. reading of all sensors (1105): 52391
```
This must be a statistical error.

This report also includes a measurement of the time taken in the rest of one iteration in the main loop, after reading from the sensors (`After reading`):
```
Avg. Mux switch (4333): 135
Avg. Buf reset (4333): 0
Avg. Buf wait (4333): 208
Avg. Buf read (4333): 1297
Avg. reading of all sensors (867): 65413
Avg. After reading (867): 17068
```


Although the improvement is minor, there was one. Now a next idea for optimization includes utilizing the [second processor](https://randomnerdtutorials.com/esp32-dual-core-arduino-ide/) in the ESP32 to reset all sensors while the readings are getting processed and the rest of the loop is executed.

**Comment by xjjak** [comment]
Using `xTaskCreatePinnedToCore`, a task can be implemented in freeRTOS. These should run continually, so the callback `task_fifo_reset` contains an infinite for loop, that waits for the flag `task_reset_fifos_flag` to be set to true.
Setting it back to `false` doesn't work yet though, due to pointer fuckery #12 .

**Comment by xjjak** [comment]
As mentioned in the resolved #12 the pointer issues were resolved by using the proper way to handle flags in tasks.

Although parallelization worked in principle, no improvements were noticeable, the measurements rather worsened. After testing with and without parallelization and `reset_all_bufs()` at the start and end of `get_all_readings()`, the readings were making no sense, at around `60ms` for *Avg. reading of all sensors*. 

The only thing left to try was to take out most of the smaller "breakpoints".  This was introduced to make en- and disabling them convenient: 
```cpp
#define __TIMING(format)   if (TIMING_ENABLED) {Serial.printf(format, micros()-prev_micros); prev_micros = micros();}
const int TIMING_ENABLED = 0;
```

After disabling all outputs except for *Avg. readings of all sensors* (`All readings`) the results even without parallelization were as surprising as astonishing, achieving `~11ms` per reading cycle. Re-enabling the execution of `reset_all_bufs()` on *Core 0* further brought this down to **~8.5ms per gathering** while the time taken for the rest of the main loop remained at about `11ms`.

In comparison to previous measurements, the recorded time between measurements in total was *~50ms*, now being around *~20ms*, which is an **improvement of ~30ms**. 

The only options for further improvement coming to mind at this point in time is improving on the rest of the main loop, moving from string output to binary, both over Serial and to SDcard.
